File: app/src/main/java/cl/figonzal/evaluatool/evalua/evalua3/Evalua3Activity.java
/*-----------------------------------------------------------------------------
                                                                              -
 This file is subject to the terms and conditions defined in                  -
 file 'LICENSE', which is part of this source code package                    -
                                                                              -
 Autor: Felipe González                                                       -
 Email: felipe.gonzalezalarcon94@gmail.com                                    -
                                                                              -
 Copyright (c) 2020                                                           -
                                                                              -
 Last modified 14-08-20 1:09                                                  -
 -----------------------------------------------------------------------------*/

package cl.figonzal.evaluatool.evalua.evalua3;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.crashlytics.FirebaseCrashlytics;

import java.util.ArrayList;
import java.util.List;

import cl.figonzal.evaluatool.R;
import cl.figonzal.evaluatool.adapter.EvaluaAdapter;
import cl.figonzal.evaluatool.interfaces.AbrirActivity;
import cl.figonzal.evaluatool.modelo.Evalua;
import cl.figonzal.evaluatool.utilidades.ConfigRoutes;
import cl.figonzal.evaluatool.utilidades.Utilidades;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionAdapter;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter;

public class Evalua3Activity extends AppCompatActivity implements EvaluaAdapter.ClickListener, AbrirActivity {

    private SectionedRecyclerViewAdapter sectionedRecyclerViewAdapter;

    private FirebaseCrashlytics crashlytics;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_evalua3);

        crashlytics = FirebaseCrashlytics.getInstance();

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitleTextColor(getResources().getColor(R.color.colorOnPrimary, getTheme()));
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        assert actionBar != null;
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp);
        actionBar.setTitle(getString(R.string.TOOLBAR_EVALUA_3));

        configurarExpandedList();
    }

    private void configurarExpandedList() {
        sectionedRecyclerViewAdapter = new SectionedRecyclerViewAdapter();

        //SUBTIMES - SUBMODULO 1
        List<Evalua> subItems1 = new ArrayList<>();
        subItems1.add(new Evalua(getString(R.string.EVALUA_3_M1_SI_1)));

        List<Evalua> subItem2 = new ArrayList<>();
        subItem2.add(new Evalua(getString(R.string.EVALUA_3_M2_SI_1)));
        subItem2.add(new Evalua(getString(R.string.EVALUA_3_M2_SI_2)));
        subItem2.add(new Evalua(getString(R.string.EVALUA_3_M2_SI_3)));
        subItem2.add(new Evalua(getString(R.string.EVALUA_3_EVALUA_GLOBAL)));

        List<Evalua> subItem3 = new ArrayList<>();
        subItem3.add(new Evalua(getString(R.string.EVALUA_3_M3_SI_1)));

        List<Evalua> subItem4 = new ArrayList<>();
        subItem4.add(new Evalua(getString(R.string.EVALUA_3_M4_SI_1)));
        subItem4.add(new Evalua(getString(R.string.EVALUA_3_M4_SI_2)));
        subItem4.add(new Evalua(getString(R.string.EVALUA_3_EVALUA_GLOBAL)));

        List<Evalua> subItem5 = new ArrayList<>();
        subItem5.add(new Evalua(getString(R.string.EVALUA_3_M5_SI_1)));
        subItem5.add(new Evalua(getString(R.string.EVALUA_3_M5_SI_2)));

        List<Evalua> subItem6 = new ArrayList<>();
        subItem6.add(new Evalua(getString(R.string.EVALUA_3_M6_SI_1)));
        subItem6.add(new Evalua(getString(R.string.EVALUA_3_M6_SI_2)));
        subItem6.add(new Evalua(getString(R.string.EVALUA_3_EVALUA_GLOBAL)));

        //HEADERS
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_3_MODULO_1), subItems1, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_3_MODULO_2), subItem2, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_3_MODULO_3), subItem3, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_3_MODULO_4), subItem4, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_3_MODULO_5), subItem5, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_3_MODULO_6), subItem6, this));

        final RecyclerView recyclerView = findViewById(R.id.recyclerview);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        recyclerView.setAdapter(sectionedRecyclerViewAdapter);
    }

    @Override
    public void onItemRootViewClicked(@NonNull String sectionTitle, int itemAdapterPosition) {

        Utilidades.handleRoutes(
                new ConfigRoutes(this).getRouteMapEvalua3(),
                sectionTitle,
                sectionedRecyclerViewAdapter,
                itemAdapterPosition,
                this
        );
    }

    @Override
    public void onHeaderRootViewClicked(@NonNull final EvaluaAdapter section) {

        SectionAdapter sectionAdapter = sectionedRecyclerViewAdapter.getAdapterForSection(section);

        final boolean wasExpanded = section.isExpanded();
        final int previousItemsTotal = section.getContentItemsTotal();

        section.setExpanded(!wasExpanded);
        sectionAdapter.notifyHeaderChanged();

        if (wasExpanded) {
            sectionAdapter.notifyItemRangeRemoved(0, previousItemsTotal);
        } else {
            sectionAdapter.notifyAllItemsInserted();
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        if (item.getItemId() == android.R.id.home) {

            Log.d(getString(R.string.CLOSE_EVALUA_3), getString(R.string.ACTIVIDAD_CERRADA));

            crashlytics.log(getString(R.string.CLOSE_EVALUA_3) + getString(R.string.ACTIVIDAD_CERRADA));

            finish();

            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void abrirActividad(Activity activity, Class<?> calledActivity, String log_title, String log_reponse) {

        Log.d(log_title, log_reponse);
        crashlytics.log("D/" + log_title + ": " + log_reponse);

        Intent intent = new Intent(activity, calledActivity);
        startActivity(intent);
    }

}

File: app/src/main/java/cl/figonzal/evaluatool/evalua/evalua3/Evalua3Activity.kt
/*

 This file is subject to the terms and conditions defined in
 file 'LICENSE', which is part of this source code package

 Autor: Felipe González
 Email: felipe.gonzalezalarcon94@gmail.com

 Copyright (c) 2021

 Last modified 07-05-21 17:32
 */
package cl.figonzal.evaluatool.evalua.evalua3

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import cl.figonzal.evaluatool.R
import cl.figonzal.evaluatool.adapter.EvaluaAdapter
import cl.figonzal.evaluatool.adapter.EvaluaAdapter.ClickListener
import cl.figonzal.evaluatool.databinding.ActivityEvalua3Binding
import cl.figonzal.evaluatool.modelo.Evalua
import cl.figonzal.evaluatool.utilidades.ConfigRoutes
import cl.figonzal.evaluatool.utilidades.RouteHandler
import cl.figonzal.evaluatool.utilidades.configActionBar
import cl.figonzal.evaluatool.utilidades.logInfo
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter
import java.util.*

class Evalua3Activity : AppCompatActivity(), ClickListener {

    private lateinit var binding: ActivityEvalua3Binding
    private var sectionedRecyclerViewAdapter = SectionedRecyclerViewAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityEvalua3Binding.inflate(layoutInflater)
        setContentView(binding.root)

        configActionBar(R.string.TOOLBAR_EVALUA_3, binding.include.toolbar)
        configurarExpandedList()
    }

    private fun configurarExpandedList() {

        //SUBTIMES - SUBMODULO 1
        val subItems1 = listOf(
            Evalua(getString(R.string.EVALUA_3_M1_SI_1))
        )

        val subItem2 = listOf(
            Evalua(getString(R.string.EVALUA_3_M2_SI_1)),
            Evalua(getString(R.string.EVALUA_3_M2_SI_2)),
            Evalua(getString(R.string.EVALUA_3_M2_SI_3)),
            Evalua(getString(R.string.EVALUA_3_VALORACION_GLOBAL_RAZONAMIENTO))
        )

        val subItem3 = listOf(
            Evalua(getString(R.string.EVALUA_3_M3_SI_1))
        )

        val subItem4 = listOf(
            Evalua(getString(R.string.EVALUA_3_M4_SI_1)),
            Evalua(getString(R.string.EVALUA_3_M4_SI_2)),
            Evalua(getString(R.string.EVALUA_3_VALORACION_GLOBAL_LECTURA))
        )

        val subItem5 = listOf(
            Evalua(getString(R.string.EVALUA_3_M5_SI_1)),
            Evalua(getString(R.string.EVALUA_3_M5_SI_2)),
            Evalua(getString(R.string.EVALUA_2_VALORACION_GLOBAL_ESCRITURA))
        )

        val subItem6 = listOf(
            Evalua(getString(R.string.EVALUA_3_M6_SI_1)),
            Evalua(getString(R.string.EVALUA_3_M6_SI_2)),
            Evalua(getString(R.string.EVALUA_3_VALORACION_GLOBAL_MATEMATICA))
        )

        //HEADERS
        sectionedRecyclerViewAdapter.apply {

            addSection(
                EvaluaAdapter(
                    getString(R.string.EVALUA_3_MODULO_1),
                    subItems1,
                    this@Evalua3Activity
                )
            )
            addSection(
                EvaluaAdapter(
                    getString(R.string.EVALUA_3_MODULO_2),
                    subItem2,
                    this@Evalua3Activity
                )
            )
            addSection(
                EvaluaAdapter(
                    getString(R.string.EVALUA_3_MODULO_3),
                    subItem3,
                    this@Evalua3Activity
                )
            )
            addSection(
                EvaluaAdapter(
                    getString(R.string.EVALUA_3_MODULO_4),
                    subItem4,
                    this@Evalua3Activity
                )
            )
            addSection(
                EvaluaAdapter(
                    getString(R.string.EVALUA_3_MODULO_5),
                    subItem5,
                    this@Evalua3Activity
                )
            )
            addSection(
                EvaluaAdapter(
                    getString(R.string.EVALUA_3_MODULO_6),
                    subItem6,
                    this@Evalua3Activity
                )
            )
        }

        //RecyclerView 3
        binding.rv3.apply {
            layoutManager = LinearLayoutManager(applicationContext)
            adapter = sectionedRecyclerViewAdapter
        }

    }

    override fun onItemRootViewClicked(sectionTitle: String, itemAdapterPosition: Int) {
        ConfigRoutes.routeMap[getString(R.string.routeMapEvalua3)]?.let {
            RouteHandler.handleRoutes(
                it,
                sectionTitle,
                sectionedRecyclerViewAdapter,
                itemAdapterPosition,
                this
            )
        }
    }

    override fun onHeaderRootViewClicked(section: EvaluaAdapter) {

        val sectionAdapter = sectionedRecyclerViewAdapter.getAdapterForSection(section)
        val wasExpanded = section.isExpanded
        val previousItemsTotal = section.contentItemsTotal

        section.isExpanded = !wasExpanded
        sectionAdapter.notifyHeaderChanged()

        when {
            wasExpanded -> sectionAdapter.notifyItemRangeRemoved(0, previousItemsTotal)
            else -> sectionAdapter.notifyAllItemsInserted()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        if (item.itemId == android.R.id.home) {
            logInfo(R.string.ACTIVIDAD_CERRADA)
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}